runtime/src/accounts_db.rs

```diff
@@ -17671,4 +17701,402 @@ pub mod tests {
+                .binary_search_by_key(&num_children.unwrap_or(0), |(size, _)| *size)
+            {
+                Ok(found_position) => found_position,
+                Err(woudbe_position) => woudbe_position,
+            };
+            test_largest_keys.insert(key_position, (num_children.unwrap_or(0), *wallet));
+            // If there was indeed more elements with the same key size sort them by Pubkey
+            if existing_key_size_position.is_some() {
+                // Obtain a slice of mutable references to all elements with the same key_size
+                let mut sub_slice = test_largest_keys
+                    .split_mut(|(k, _)| *k != num_children.unwrap_or(0))
+                    .flatten()
+                    .collect_vec();
+                // Sort them...
+                let mut sorting_buffer = sub_slice.iter().map(|x| *(*x)).collect_vec();
+                sorting_buffer.sort_unstable_by_key(|(_, v)| *v);
+                // Copy back into the list
+                for i in 0..sub_slice.len() {
+                    *(sub_slice[i]) = (sorting_buffer[i].0, sorting_buffer[i].1);
+                }
+            }
+            // Prune list
+            while test_largest_keys.len() > MAX_NUM_LARGEST_INDEX_KEYS_RETURNED {
+                test_largest_keys.remove(0);
+            }
+        }
+
+        // Verify secondary index list matches the list you tracked above.
+        let largest_keys = accounts_db.accounts_index.get_largest_keys(
+            &AccountIndex::ProgramId,
+            MAX_NUM_LARGEST_INDEX_KEYS_RETURNED,
+        );
+        // Reverse the tracking vector and check for equality
+        test_largest_keys = test_largest_keys.into_iter().rev().collect_vec();
+        assert_eq!(test_largest_keys, largest_keys);
+
+        // Test querries for a partial list and past max return size
+        for i in 0..=MAX_NUM_LARGEST_INDEX_KEYS_RETURNED + 1 {
+            test_largest_keys = accounts_db
+                .accounts_index
+                .get_largest_keys(&AccountIndex::ProgramId, i);
+            if i <= MAX_NUM_LARGEST_INDEX_KEYS_RETURNED {
+                assert_eq!(test_largest_keys.len(), i);
+            } else {
+                assert_eq!(test_largest_keys.len(), MAX_NUM_LARGEST_INDEX_KEYS_RETURNED);
+            }
+        }
+
+        let num_token_accounts_range = Uniform::new_inclusive(1, MAX_TOKEN_ACCOUNTS);
+        let num_mint_accounts_range = Uniform::new_inclusive(NUM_WALLETS, MAX_MINT_ACCOUNTS);
+        let wallet_pubkey_index_range = Uniform::new(0, NUM_WALLETS);
+
+        let num_token_accounts = (&mut num_generator)
+            .sample_iter(num_token_accounts_range)
+            .next();
+        let num_mint_accounts = (&mut num_generator)
+            .sample_iter(num_mint_accounts_range)
+            .next();
+
+        // Make a bunch of SPL Tokens each with some random number of SPL Token Accounts that have the token in them
```

Comment:
The addition is to the SecondaryIndex type which is generic between all of them. We're testing from AccountsDB rather than AccountsIndex because we need to test removal at the end which required access to the bank functions to root and purge a slot. Thus the only interface between them that is different is AccountsIndex::get_largest_keys() which is tested over all 3 in the AdminRPC test. I can add a simple test over the interface here if you'd like, but the logic doesn't need to be duplicated differently for the other secondary indexes because they are not of a different type.

Reply:
Um, what? Why bother adding any token accounts at all if you're not doing something like this at some point(s):
```
let largest_keys = accounts_db.accounts_index.get_largest_keys(
    &AccountIndex::SplTokenMint,
    MAX_NUM_LARGEST_INDEX_KEYS_RETURNED,
);
assert_eq!(largest_keys, <EXPECTED_LARGEST_KEYS_LIST>);

let largest_keys = accounts_db.accounts_index.get_largest_keys(
    &AccountIndex::SplTokenOwner,
    MAX_NUM_LARGEST_INDEX_KEYS_RETURNED,
);
assert_eq!(largest_keys, <EXPECTED_LARGEST_KEYS_LIST>);
```

Sure, those accounts also add a bunch to the AccountIndex::ProgramId spl-token program-id key, but you've already tested that behavior with the "child" accounts.

Reply:
Hmm, one single token account would be enough to check for non-empty. I don't think all the current code bloat is justified, as it makes the test harder to read and understand. My preference at this point is to strip out all the token-related code (will save 100+ lines!)